### PR TITLE
[DAT-16941] - we should keep db-natural order for StoLo objects

### PR DESCRIPTION
**Pull request by v-petrovych**

### CODE REVIEW

File: liquibase-core/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java [hunk 1]
```diff
@@ -245,12 +245,8 @@ public int compare(DatabaseObject o1, DatabaseObject o2) {
                     if (i != 0) {
                         return i;
                     }
-                } else if (o1 instanceof StoredDatabaseLogic && o1.getAttribute("order", Integer.class) != null
```
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java:248]: Not sure that returning here 1 to keep ordering form the LinkedHashSet is a good idea - we try to break design by doing it, which means that we try to go in a wrong way

File: liquibase-core/src/main/java/liquibase/structure/DatabaseObjectCollection.java [hunk 2]
```diff
@@ -75,7 +74,7 @@ public void add(DatabaseObject databaseObject) {
         }
         Map<String, Set<DatabaseObject>> collectionMap = cache.get(databaseObject.getClass());
         if (collectionMap == null) {
-            collectionMap = new HashMap<String, Set<DatabaseObject>>();
+            collectionMap = new LinkedHashMap<>();
```
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/structure/DatabaseObjectCollection.java:77]: If I understand correctly key of this LinkedHashMap is object name, so order for these objects is going to be preserved by object type, why do we need this change? i.e. I can't fully understand how it helps to preserve specific objects ordering, i.e. that some_func_1 is going before some_func_2

File: liquibase-core/src/main/java/liquibase/structure/core/Schema.java [hunk 3]
```diff
@@ -142,7 +142,7 @@ public void addDatabaseObject(DatabaseObject databaseObject) {
         }
         Set<DatabaseObject> objects = this.getObjects().get(databaseObject.getClass());
         if (objects == null) {
-            objects = new HashSet<DatabaseObject>();
+            objects = new LinkedHashSet<DatabaseObject>();
```
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/structure/core/Schema.java:145]: Same comment about this LinkedHashSet as above

File: liquibase-core/src/main/java/liquibase/diff/DiffResult.java [hunk 4]
```diff
@@ -69,7 +68,7 @@ public Set<? extends DatabaseObject> getMissingObjects() {
     }
 
     public <T extends DatabaseObject> Set<T> getMissingObjects(Class<T> type) {
-        Set returnSet = new HashSet();
+        Set returnSet = new LinkedHashSet();
```
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/diff/DiffResult.java:71]: I see that tests pass, but I don't understand how these changes make it :) i.e. even this part - where HashSet is changed to LinkedHashSet to preserve order looks strange as we anyway add values into it from missingObjects, which is also a HashSet, so even if we didn't use LinkedHashSet - shouldn't we just get the same order anyway?

**Review by andrulban — COMMENTED** [review]
Hi @v-petrovych, left a few comments just to be able to understand the solution better as currently I don't see a full picture how all these changes end up with doing what they are doing (most probably will need a sync to be able to get the idea)

File: liquibase-core/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java [hunk 5]
```diff
@@ -245,12 +245,8 @@ public int compare(DatabaseObject o1, DatabaseObject o2) {
...
-                        && o2.getAttribute("order", Integer.class) != null) {
-                    int order = o1.getAttribute("order", Integer.class).compareTo(o2.getAttribute("order", Integer.class));
-                    if (order != 0) {
-                        return order;
-                    }
+                } else if (o1 instanceof StoredDatabaseLogic) {
+                    return 1;
```
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java]: Let's add a comment why we made this change

File: liquibase-core/src/main/java/liquibase/diff/DiffResult.java (hunk 4 above)
Comment (andrulban) [inline liquibase-core/src/main/java/liquibase/diff/DiffResult.java:71]: Do we need to make similar changes to changedObjects and unexpectedObjects?

**Review by andrulban — COMMENTED** [review]
Overall changes look like adding workarounds, but I see no other way how we can fix the bug, as PostgreSQL does not see dependencies between functions and procedures so we can't build a proper dependency graph...

Let's proceed, but let's add @wwillard7800 @StevenMassaro and @suryaaki2 to make additional reviews, because there should be a similar issue in foundation repo (with depending on each other functions and procedures for Postgresql), so it might be useful and also we might get any other ideas if there is any other way to fix it....

**Comment by v-petrovych** [comment]
Hi Team, @wwillard7800 @StevenMassaro and @suryaaki2, please review these changes when you'll have time and leave your comments/concerns if any. Many thanks!

P.S.
Short context of these changes: Postgres doesn't allow us neither to track dependencies between routines (procedures or functions) nor to order them by 'datatime_created' or so. The only way to avoid issues during generated changelog deployment is to keep exactly the same order for routines as it is in 'pg_proc' table. So the sense of these changes is 1) use ordered LinkedHash collections in particular places 2) return 1 in Comparator if db object is StoLo (consider pair o1 and o2 always "already ordered")

**Comment by StevenMassaro** [comment]
I created a new ticket for the pro dev team to look into whether these changes apply to Liquibase as well. https://datical.atlassian.net/browse/DAT-17937

**Review by andrulban — APPROVED** [review]
Approving taking into account that foundation team also took a look and didn't have any objections

**Pull request merged**